paperqa/agents/docs.py

```diff
@@ -0,0 +1,250 @@
+    "mistralai/Mixtral-8x7B-Instruct-v0.1": (1.0 / 10**6, 1.0 / 10**6),
+    "mistralai/Mixtral-8x22B-Instruct-v0.1": (1.0 / 10**6, 1.0 / 10**6),
+}
+
+
+def compute_model_token_cost(model: str, tokens: int, is_completion: bool) -> float:
+    if model in costs:  # Prefer our internal costs model
+        model_costs: tuple[float, float] = costs[model]
+    else:
+        logger.warning(f"Model {model} not found in costs.")
+        return 0.0
+    return tokens * model_costs[int(is_completion)]
+
+
+def compute_total_model_token_cost(token_counts: dict[str, list[int]]) -> float:
+    """Sum the token counts for each model and return the total cost."""
+    cost = 0.0
+    for model, tokens in token_counts.items():
+        if sum(tokens) > 0:
+            cost += compute_model_token_cost(
+                model, tokens=tokens[0], is_completion=False
+            ) + compute_model_token_cost(model, tokens=tokens[1], is_completion=True)
+    return cost
+
+
+async def stream_filter(
+    docs: Docs,
+    query: str,
+    answer: Answer,
+    num_doc_match_retries: int = 1,
+    adoc_match_threshold: int = QueryRequest.model_fields[
+        "adoc_match_threshold"
+    ].default,
+) -> Answer:
+    """Set the `answer.dockey_filter`."""
+    if len(docs.docs) >= adoc_match_threshold:
+
+        for i in range(num_doc_match_retries + 1):
+            try:
+                result = await docs.adoc_match(
+                    query,
+                    rerank=True,  # want to set it explicitly
+                    answer=answer,
+                )
+                break
+            except Exception as e:
+                if i < num_doc_match_retries:
+                    logger.warning(
+                        f"Failed to filter, possibly due to huge paper citation, full message: {e}."
+                    )
+                    continue
+                raise
+    else:
+        result = set(docs.docs.keys())
+    answer.dockey_filter = result
+    return answer
+
+
+async def stream_evidence(
+    docs: Docs,
+    request: QueryRequest,
+    answer: Answer | None = None,
+) -> Answer:
+    if answer is None:
+        answer = await stream_filter(
+            docs=docs,
+            query=request.query,
+            answer=Answer(question=request.query, id=request.id),
+            adoc_match_threshold=request.adoc_match_threshold,
+        )
+
+    # clear texts
+    docs.jit_texts_index = True
+    # ensure length is set correctly
+    answer.summary_length = request.summary_length
+    answer = await docs.aget_evidence(
+        answer=answer,
+        max_sources=request.max_sources,
+        k=request.consider_sources,
+        detailed_citations=True,
+    )
+    # we uniquify the contexts
+    names = set()
+    uniq_docs = []
+    for c in answer.contexts:
+        if c.text.doc.docname not in names:
+            uniq_docs.append(c.text.doc)
+            names.add(c.text.doc.docname)
+    return answer
+
+
+async def stream_cost(
```

Comment:
If possible, can we get rid of all of these stream functions entirely? They are just added indirection imo

Reply:
ok -- the whole file could actually be refactored out, much better here now.